orchagent/bfdorch.cpp

```diff
@@ -0,0 +1,412 @@
+    uint32_t tx_interval = BFD_SESSION_DEFAULT_TX_INTERVAL;
+    uint32_t rx_interval = BFD_SESSION_DEFAULT_RX_INTERVAL;
+    uint8_t multiplier = BFD_SESSION_DEFAULT_DETECT_MULTIPLIER;
+    bool multihop = false;
+    MacAddress dst_mac;
+    bool dst_mac_provided = false;
+    bool src_ip_provided = false;
+
+    sai_attribute_t attr;
+    vector<sai_attribute_t> attrs;
+    vector<FieldValueTuple> fvVector;
+
+    for (auto i : data)
+    {
+        auto value = fvValue(i);
+
+        if (fvField(i) == "tx_interval")
+            tx_interval = to_uint<uint32_t>(value);
+        else if (fvField(i) == "rx_interval")
+            rx_interval = to_uint<uint32_t>(value);
+        else if (fvField(i) == "multiplier")
+            multiplier = to_uint<uint8_t>(value);
+        else if (fvField(i) == "multihop")
+            multihop = (value == "true") ? true : false;
+        else if (fvField(i) == "local_addr")
+        {
+            src_ip = IpAddress(value);
+            src_ip_provided = true;
+        }
+        else if (fvField(i) == "type")
+        {
+            if (session_type_map.find(value) == session_type_map.end())
+            {
+                SWSS_LOG_ERROR("Invalid BFD session type %s\n", value.c_str());
+                continue;
+            }
+            bfd_session_type = session_type_map.at(value);
+        }
+        else if (fvField(i) == "dst_mac")
+        {
+            dst_mac = MacAddress(value);
+            dst_mac_provided = true;
+        }
+        else
+            SWSS_LOG_ERROR("Unsupported BFD attribute %s\n", fvField(i).c_str());
+    }
+
+    attr.id = SAI_BFD_SESSION_ATTR_TYPE;
+    attr.value.s32 = bfd_session_type;
+    attrs.emplace_back(attr);
+    fvVector.emplace_back("type", session_type_loopup.at(bfd_session_type));
+
+    attr.id = SAI_BFD_SESSION_ATTR_LOCAL_DISCRIMINATOR;
+    attr.value.u32 = bfd_gen_id();
+    attrs.emplace_back(attr);
+
+    attr.id = SAI_BFD_SESSION_ATTR_UDP_SRC_PORT;
+    attr.value.u32 = bfd_src_port();
+    attrs.emplace_back(attr);
+
+    attr.id = SAI_BFD_SESSION_ATTR_REMOTE_DISCRIMINATOR;
+    attr.value.u32 = 0;
+    attrs.emplace_back(attr);
+
+    attr.id = SAI_BFD_SESSION_ATTR_BFD_ENCAPSULATION_TYPE;
+    attr.value.s32 = encapsulation_type;
+    attrs.emplace_back(attr);
+
+    attr.id = SAI_BFD_SESSION_ATTR_IPHDR_VERSION;
+    attr.value.u8 = src_ip.isV4() ? 4 : 6;
+    attrs.emplace_back(attr);
+
+    if (!src_ip_provided)
+    {
+        SWSS_LOG_ERROR("Failed to create BFD session %s because source IP is not provided", key.c_str());
+        return false;
+    }
+    attr.id = SAI_BFD_SESSION_ATTR_SRC_IP_ADDRESS;
+    copy(attr.value.ipaddr, src_ip);
+    attrs.emplace_back(attr);
+    fvVector.emplace_back("local_addr", src_ip.to_string());
+
+    attr.id = SAI_BFD_SESSION_ATTR_DST_IP_ADDRESS;
+    copy(attr.value.ipaddr, peer_address);
+    attrs.emplace_back(attr);
+
+    attr.id = SAI_BFD_SESSION_ATTR_MIN_TX;
+    attr.value.u32 = tx_interval * 1000;
```

Comment:
define this. don't use magic number

Reply:
Got this defined as `BFD_SESSION_MILLISECOND_TO_MICROSECOND`